### PR TITLE
enhance jit

### PR DESCRIPTION
**Pull request by tensor-tang**
enhance and complete #14783 

对底层的函数，包括：
- 包括vmul, vaddrelu,等基本函数，
- relu , exp, sigmoid, tanh 等激活类。
- lstm，gru等rnn函数
- crf decoding 和layer norm等具体功能

实现了重构，xbyak，mkl，mkldnn，openblas 等第三方库都具有各自的实现，调用时根据优先级和条件自动选择。

可以参考readme https://github.com/tensor-tang/Paddle/blob/refine/jit/paddle/fluid/operators/jit/README.md

### CODE REVIEW

**Comment by tensor-tang** [comment]
PR_CI error:

>  echo 'You must have panyx0718 approval for the const_cast'

@panyx0718 这个const_cast是这里调用到的 

https://github.com/tensor-tang/Paddle/blob/a369c8025bec70dab44c1d22b09b53882b1b552f/paddle/fluid/operators/jit/gen_base.h#L33-L40

本来是这样的 https://github.com/tensor-tang/Paddle/commit/d538513fce27c9442553b39a1cff9e482c922196#diff-1fe2196680d5832de4d8573c42d2ffc2L39

反正返回函数指针，索性去掉const，不影响使用，并且解决当时的mac编译问题。

File: paddle/fluid/operators/jit/benchmark.cc [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+#include "glog/logging.h"
+#include "paddle/fluid/operators/jit/kernels.h"
+#include "paddle/fluid/platform/place.h"
+#include "paddle/fluid/platform/port.h"
+
+DEFINE_int32(burning, 10, "Burning times.");
+DEFINE_int32(repeat, 3000, "Repeat times.");
+DEFINE_int32(max_size, 1000, "The Max size would be tested.");
+
+inline double GetCurrentUS() {
```
Comment (panyx0718) [inline paddle/fluid/operators/jit/benchmark.cc]: I believe the same function exist already? Reuse it?

File: paddle/fluid/operators/jit/benchmark.cc [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+  struct timeval time;
+  gettimeofday(&time, NULL);
+  return 1e+6 * time.tv_sec + time.tv_usec;
+}
+
+template <typename T>
+void RandomVec(const int n, T* a, const T lower = static_cast<T>(-20.f),
+               const T upper = static_cast<T>(20.f)) {
+  static unsigned int seed = 100;
```
Comment (panyx0718) [inline paddle/fluid/operators/jit/benchmark.cc]: Can seed be passed in?

File: paddle/fluid/operators/jit/gen/jitcode.h [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+using reg64_t = const Xbyak::Reg64;
+using reg32_t = const Xbyak::Reg32;
+using xmm_t = const Xbyak::Xmm;
+using ymm_t = const Xbyak::Ymm;
+using zmm_t = const Xbyak::Zmm;
+using Label = Xbyak::Label;
+
+typedef enum {
+  mul = 0,
```
Comment (panyx0718) [inline paddle/fluid/operators/jit/gen/jitcode.h]: code style: Upper case

File: paddle/fluid/operators/jit/more/intrinsic/crf_decoding.cc [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+#include "paddle/fluid/operators/jit/registry.h"
+#include "paddle/fluid/platform/cpu_info.h"
+
+namespace paddle {
+namespace operators {
+namespace jit {
+namespace more {
+namespace intrinsic {
+// Note: intrinsic code is not runtime build.
+// For example, if you build code on AVX, and run on AVX512 it can only use AVX
```
Comment (panyx0718) [inline paddle/fluid/operators/jit/more/intrinsic/crf_decoding.cc:26]: so this is not JIT?

File: paddle/fluid/operators/jit/more/intrinsic/crf_decoding.cc [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+void CRFDecoding(const int seq_len, const float* x, const float* w,
+                 float* alpha, int* track, int tag_num) {
+#ifdef __AVX512F__
```
Comment (panyx0718) [inline paddle/fluid/operators/jit/more/intrinsic/crf_decoding.cc:30]: should this be decided at runtime to be jit?

File: paddle/fluid/operators/jit/benchmark.cc (hunk 1 above)
Comment (tensor-tang) [reply paddle/fluid/operators/jit/benchmark.cc]: Yes sure, I can unify one.

File: paddle/fluid/operators/jit/benchmark.cc (hunk 2 above)
Comment (tensor-tang) [reply paddle/fluid/operators/jit/benchmark.cc]: Sure

File: paddle/fluid/operators/jit/gen/jitcode.h (hunk 3 above)
Comment (tensor-tang) [reply paddle/fluid/operators/jit/gen/jitcode.h]: Sure

File: paddle/fluid/operators/jit/more/intrinsic/crf_decoding.cc (hunk 4 above)
Comment (tensor-tang) [reply paddle/fluid/operators/jit/more/intrinsic/crf_decoding.cc:26]: Yes, this just move from original implementation,  changing to jitcode should need some time.

File: paddle/fluid/operators/jit/more/intrinsic/crf_decoding.cc (hunk 5 above)
Comment (tensor-tang) [reply paddle/fluid/operators/jit/more/intrinsic/crf_decoding.cc:30]: Unfortunately, this kind intrinsic code do not support runtime, so I add this note here https://github.com/PaddlePaddle/Paddle/pull/14958/files#diff-9f5c378fa23311e48a592bd9c7a82954R25

If need support runtime, crf decoding and layer norm intrinsic codes are optimized by Intel, maybe should be changed to jitcode by them.

**Comment by tensor-tang** [comment]
py35 遇到一次 429 - test_dist_se_resnext (Failed)

File: paddle/fluid/operators/jit/README.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
+# JIT Kernel
+
+结合函数模板和JIT生成需要的kernel函数。
```
Comment (panyx0718) [inline paddle/fluid/operators/jit/README.md:3]: follow up with eng doc?

File: paddle/fluid/operators/jit/gen_base.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  virtual const char* name() const = 0;
+  virtual size_t getSize() const = 0;
+  virtual const unsigned char* getCodeInternal() = 0;
+  template <typename Func>
+  Func getCode() {
+    const unsigned char* code = this->getCodeInternal();
+    if (FLAGS_dump_jitcode) {
+      this->dumpCode(code);
+    }
+    return reinterpret_cast<Func>(const_cast<unsigned char*>(code));
```
Comment (panyx0718) [inline paddle/fluid/operators/jit/gen_base.h:39]: add a comment to explain why const_cast is necessary here in follow up.

**Review by panyx0718 — APPROVED** [review]

File: paddle/fluid/operators/jit/README.md (hunk 6 above)
Comment (tensor-tang) [reply paddle/fluid/operators/jit/README.md:3]: Sure, will address one ASAP.

File: paddle/fluid/operators/jit/gen_base.h (hunk 7 above)
Comment (tensor-tang) [reply paddle/fluid/operators/jit/gen_base.h:39]: Sure.

**Comment by tensor-tang** [comment]
ner test on V3 : 0.214ms => 0.153 ms 提升40%

```
Before : 
====== batch_size: 1, repeat: 1, threads: 1, thread id: 0, latency: 8.61721ms, fps: 116.047 ======
I1221 07:41:58.924633 19583 tester_helper.h:191] Run 5000 times...
I1221 07:41:59.996160 19583 helper.h:229] ====== batch_size: 1, repeat: 5000, threads: 1, thread id: 0, latency: 0.214303ms, fps: 4666.3 ======

After:
I1221 07:51:06.642580   426 helper.h:229] ====== batch_size: 1, repeat: 1, threads: 1, thread id: 0, latency: 2.70372ms, fps: 369.861 ======
I1221 07:51:06.642624   426 tester_helper.h:191] Run 5000 times...
I1221 07:51:07.407107   426 helper.h:229] ====== batch_size: 1, repeat: 5000, threads: 1, thread id: 0, latency: 0.152893ms, fps: 6540.51 ======
```

**Pull request merged**